src/main/java/org/logstash/tcp/SslContextBuilder.java

```diff
@@ -0,0 +1,186 @@
+        } else if (obj instanceof PKCS8EncryptedPrivateKeyInfo) { // encrypted pkcs#8
+            InputDecryptorProvider decryptor = new JceOpenSSLPKCS8DecryptorProviderBuilder().build(keyPassword);
+            PrivateKeyInfo keyInfo = ((PKCS8EncryptedPrivateKeyInfo) obj).decryptPrivateKeyInfo(decryptor);
+            privateKey = converter.getPrivateKey(keyInfo);
+        } else {
+            throw new IllegalArgumentException("unexpected key format (" + (obj == null ? null : obj.getClass()) + ")");
+        }
+
+        for (String certPath : extraChainCerts) {
+            certChain.addAll( getCertificatesFromFile(certPath, certFactory) );
+        }
+
+        io.netty.handler.ssl.SslContextBuilder sslContextBuilder =
+                io.netty.handler.ssl.SslContextBuilder.forServer(privateKey,
+                        keyPassword == null ? null : new String(keyPassword),
+                        certChain.toArray(new X509Certificate[certChain.size()])
+                );
+
+        List<Certificate> trustedCerts = new ArrayList<>();
+        for (String certPath : certificateAuthorities) {
+            trustedCerts.addAll( getCertificatesFromFile(certPath, certFactory) );
+        }
+
+        if (!trustedCerts.isEmpty()) {
+            sslContextBuilder.trustManager(trustedCerts.toArray(new X509Certificate[trustedCerts.size()]));
+        }
+
+        sslContextBuilder.clientAuth(shouldVerify ? ClientAuth.REQUIRE : ClientAuth.NONE);
+
+        try {
+            return sslContextBuilder.build();
+        } catch (SSLException e) {
+            logger.debug("Failed to initialize SSL", e);
+            // unwrap generic wrapped exception from Netty's JdkSsl{Client|Server}Context
+            if ("failed to initialize the server-side SSL context".equals(e.getMessage()) ||
+                "failed to initialize the client-side SSL context".equals(e.getMessage())) {
+                // Netty catches Exception and simply wraps: throw new SSLException("...", e);
+                if (e.getCause() instanceof Exception) throw (Exception) e.getCause();
+            }
+            throw e;
+        } catch (Exception e) {
+            logger.debug("Failed to initialize SSL", e);
+            throw e;
+        }
+    }
+
+    private static List<Certificate> getCertificatesFromFile(final String file, final CertificateFactory factory)
+        throws IOException, CertificateException {
+        final ArrayList<Certificate> certificates = new ArrayList<>();
+        final FileInputStream fis = new FileInputStream(file);
+        try {
+            while (fis.available() > 0) {
+                try {
+                    certificates.add( factory.generateCertificate(fis) );
+                } catch (CertificateException ex) {
+                    Throwable cause = ex.getCause();
+                    if (cause != null && "Empty input".equals(cause.getMessage())) {
+                        continue;
```

Comment:
Is there a case for `debug` or `trace` logging upon receiving this error, or is it always benign?

Reply:
good call, I simply converted from the .rb parts and did not revisit with potential logging.
the whole logic seems a bit unusual - would rather just use `generateCertificates` to read multiple entries.
anyhow, let's keep that as before and I am going to add a debug line around the caught "empty" exception.

Reply:
added debug logging + an extra check when expecting to read at least one certificate from `ssl_cert`
also worth noting (in the changelog) we no longer have the side effect of setting up the BC provider globally.